File: assets/Entity/Tetromino/Tetromino.ts
import { Vec2 } from "cc";

export class Tetromino {
    //4x4的方塊，0為空
    public shape: number[][] = []
    //方塊的位置，LeftTop為(0,0)
    public position: Vec2 = new Vec2(0, 0);
    //方塊的盤面，0為空
    private board: number[][] = [];


    public constructor(shape: number[][], board: number[][]) {

        if (board.length < 10 || board[0].length < 10) {
            throw new Error("Board must be at least 10x10");
        }
        this.shape = shape;
        this.board = board;
    }



    //嘗試旋轉方塊，如果成功則旋轉並回傳true，否則回傳false
    public tryRotateL(): boolean {
        return this.tryRotate(false);
    }

    //嘗試旋轉方塊，如果成功則旋轉並回傳true，否則回傳false
    public tryRotateR(): boolean {
        return this.tryRotate(true);
    }

    //嘗試移動方塊，如果成功則移動並回傳true，否則回傳false
    public tryMove(vec: Vec2): boolean {
        let nextPosition = this.position.clone().add(vec);
        if (this.isLegal(this.shape, nextPosition)) {
            this.position = nextPosition;
            return true;
        }
        return false;
    }

    private rotate(matrix: number[][],clockwise:boolean): void {
        const n: number = matrix.length;
        let width: number = n - 1;
    
        for (let i = 0; i < Math.floor(n / 2); i++) {
            for (let j = i; j < i + width; j++) {
                if (clockwise) {
                    const temp: number = matrix[i][j];
                    matrix[i][j] = matrix[n - j - 1][i];
                    matrix[n - j - 1][i] = matrix[n - i - 1][n - j - 1];
                    matrix[n - i - 1][n - j - 1] = matrix[j][n - i - 1];
                    matrix[j][n - i - 1] = temp;
                } else {
                    const temp: number = matrix[i][j];
                    matrix[i][j] = matrix[j][n - i - 1];
                    matrix[j][n - i - 1] = matrix[n - i - 1][n - j - 1];
                    matrix[n - i - 1][n - j - 1] = matrix[n - j - 1][i];
                    matrix[n - j - 1][i] = temp;
                }
            }
            width -= 2;
        }
    }

    //嘗試旋轉方塊，如果失敗則進行WallKick檢查，如果還是失敗則不旋轉
    private tryRotate(clockwise: boolean): boolean {

        let nextShape = this.shape.map(inner => [...inner]);
        this.rotate(nextShape,clockwise);
        if (this.isLegal(nextShape, this.position)) {
            this.shape = nextShape;
            return true;
        } else {
            if (this.wallKick(nextShape)) {
                return true;
            }
        }
        return false;
    }



    //WallKick檢查，如果可以則進行WallKick，規則為最寬鬆且最符合直覺的規則
    private wallKick(nextShape: number[][]): boolean {
        let wallKickData: Vec2[][] = [
            [new Vec2(0, 0), new Vec2(-1, 0), new Vec2(-1, 1), new Vec2(0, -2), new Vec2(-1, -2)],
            [new Vec2(0, 0), new Vec2(1, 0), new Vec2(1, -1), new Vec2(0, 2), new Vec2(1, 2)],
            [new Vec2(0, 0), new Vec2(1, 0), new Vec2(1, 1), new Vec2(0, -2), new Vec2(1, -2)],
            [new Vec2(0, 0), new Vec2(-1, 0), new Vec2(-1, -1), new Vec2(0, 2), new Vec2(-1, 2)],
        ];
        for (let i = 0; i < wallKickData.length; i++) {
            for (let j = 0; j < wallKickData[i].length; j++) {
                let wallKick = wallKickData[i][j];
                let nextPosition = this.position.clone().add(wallKick);
                if (this.isLegal(nextShape, nextPosition)) {
                    this.position = nextPosition;
                    this.shape = nextShape;
                    return true;
                }
            }
        }
        return false;
    }


    //檢查移動是否合法(沒有與其他方塊重疊，沒有超出邊界)
    public isLegal(shape: number[][], position: Vec2): boolean {
        const boardWidth = this.board[0].length;
        const boardHeight = this.board.length;

        for (let y = 0; y < shape.length; y++) {
            for (let x = 0; x < shape[y].length; x++) {
                if (shape[y][x] == 0)
                    continue;
                const newX = position.x + x;
                const newY = position.y + y;
                if (newX < 0 || newX >= boardWidth || newY < 0 || newY >= boardHeight)
                    return false;
                if (this.board[newY][newX] !== 0)
                    return false;
            }
        }

        return true;
    }

    //快速掉落方塊
    public hardDrop(): void {
        while (this.tryMove(new Vec2(0, 1))) {
        }
    }
}

File: Assets/Tetris/Tetromino.cs
using UnityEngine;

public class Tetromino
{
    public int[,] Shape;
    public Vector2Int Position = Vector2Int.zero;
    public string Alias = "";

    public Tetromino(int[,] shape, string Alias = "")
    {
        this.Alias = Alias;
        Shape = new int[shape.GetLength(0), shape.GetLength(1)];
        for (int i = 0; i < shape.GetLength(0); i++)
        {
            for (int j = 0; j < shape.GetLength(1); j++)
            {
                Shape[i, j] = shape[i, j];
            }
        }
    }

    public Tetromino(Tetromino clone) : this(clone.Shape)
    {
        Position = new Vector2Int(clone.Position.x, clone.Position.y);
        Alias = clone.Alias;
    }


    /// <summary>
    /// 嘗試旋轉, 成功則回傳True, 失敗回傳False
    /// </summary>
    /// <param name="matrix">棋盤</param>
    /// <param name="clockwise">True -> 順時針旋轉, False 逆時針旋轉</param>
    /// <returns></returns>
    public bool TryRotate(in int[][] matrix, bool clockwise)
    {
        //複製一個暫時性的方塊用來測試旋轉是否合法
        Tetromino nextShape = new Tetromino(Shape);
        nextShape.Shape = Rotate(nextShape.Shape, clockwise);

        if (nextShape.IsLegal(matrix, Position))
        {
            Shape = nextShape.Shape;
            return true;
        }
        //TODO: 待驗證
        else if (WallKick(matrix, nextShape.Shape))
        {
            return true;
        }
        return false;
    }

    /// <summary>
    /// 嘗試移動, 如果有移動成功回傳True
    /// </summary>
    /// <param name="matrix"></param>
    /// <param name="vec"></param>
    /// <returns></returns>
    public bool TryMove(in int[][] matrix, Vector2Int vec)
    {
        Vector2Int nextPosition = Position + vec;
        if (IsLegal(matrix, nextPosition))
        {
            Position = nextPosition;
            return true;
        }
        return false;
    }

    /// <summary>
    /// 旋轉並回傳新的矩陣
    /// </summary>
    /// <param name="shape">傳入形狀矩陣</param>
    /// <param name="clockwise">T -> 順時針, F -> 逆時針</param>
    /// <returns></returns>
    private int[,] Rotate(int[,] shape, bool clockwise)
    {
        int[,] newMatrix = (int[,])shape.Clone();
        int n = newMatrix.GetLength(0);
        for (int i = 0; i < n / 2; i++)
        {
            for (int j = i; j < n - i - 1; j++)
            {
                if (clockwise)
                {
                    int temp = newMatrix[i, j];
                    newMatrix[i, j] = newMatrix[n - j - 1, i];
                    newMatrix[n - j - 1, i] = newMatrix[n - i - 1, n - j - 1];
                    newMatrix[n - i - 1, n - j - 1] = newMatrix[j, n - i - 1];
                    newMatrix[j, n - i - 1] = temp;
                }
                else
                {
                    int temp = newMatrix[i, j];
                    newMatrix[i, j] = newMatrix[j, n - i - 1];
                    newMatrix[j, n - i - 1] = newMatrix[n - i - 1, n - j - 1];
                    newMatrix[n - i - 1, n - j - 1] = newMatrix[n - j - 1, i];
                    newMatrix[n - j - 1, i] = temp;
                }
            }
        }
        return newMatrix;
    }

    /// <summary>
    /// 快速下落
    /// </summary>
    /// <param name="board"></param>
    public void HardDrop(in int[][] board)
    {
        while (IsLegal(board, Position + Vector2Int.down))
        {
            Position += Vector2Int.down;
        }
    }

    /// <summary>
    /// 牆踢(直接編輯)
    /// </summary>
    /// <param name="board"></param>
    /// <param name="nextShape">形狀</param>
    /// <returns>成功與否</returns>
    private bool WallKick(in int[][] board, int[,] nextShape)
    {
        Vector2Int[] wallKickTests = {
            new Vector2Int(-1, 0),
            new Vector2Int(1, 0),
            new Vector2Int(0, -1),
            new Vector2Int(0, 1),
            new Vector2Int(-1, -1),
            new Vector2Int(1, -1),
            new Vector2Int(-1, 1),
            new Vector2Int(1, 1)
        };

        foreach (var test in wallKickTests)
        {
            if (IsLegal(board, Position + test))
            {
                Shape = nextShape;
                Position += test;
                return true;
            }
        }
        return false;
    }

    /// <summary>
    /// 判斷這個方塊是否可以放在特定位置
    /// </summary>
    /// <param name="board"></param>
    /// <param name="position"></param>
    /// <returns></returns>
    public bool IsLegal(in int[][] board, Vector2Int position)
    {
        for (int y = 0; y < Shape.GetLength(0); y++)
        {
            for (int x = 0; x < Shape.GetLength(1); x++)
            {
                if (Shape[y, x] != 0)
                {
                    int boardX = position.x + x;
                    int boardY = position.y + y;
                    if (boardX < 0 || boardX >= board[0].Length
                     || boardY < 0 || boardY >= board.Length
                     || board[boardY][boardX] != 0)
                    {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}
